README.md
# glassy

A Flutter package for all platforms which provides Glass Effects

## Screenshots

<img src="https://github.com/smae1993/glassy/raw/main/screenShots/screen_shot_2.png" height="400em" width="225em" />

## Usage

[Example](https://github.com/smae1993/glassy/blob/main/example/lib/main.dart)

To use this package :

* add the dependency to your [pubspec.yaml](https://github.com/smae1993/glassy/blob/main/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    glassy: latest version
```

### How to use

```dart
void main() {
  /// This Config is Optional
  Glassy().setConfig(GlassyConfig(
      radius: 15,
      backgroundColor: Colors.grey,
      backgroundOpacity: 0.2,
      borderOpacity: 0.5));
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Glassy Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const DemoPage(),
    );
  }
}

class DemoPage extends StatelessWidget {
  const DemoPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        fit: StackFit.expand,
        children: [
          Image.asset("assets/images/test.jpg", fit: BoxFit.cover),
          SizedBox(
            child: Column(children: [
              GlassyCard(
                  child: Row(
                children: [
                  IconButton(
                      onPressed: () {},
                      icon: const Icon(
                        Icons.search,
                        color: Colors.white,
                      )),
                  const Expanded(
                      child: Text(
                    "This is a GlassyCard",
                    textAlign: TextAlign.center,
                    style: TextStyle(
                      color: Colors.white,
                    ),
                  ))
                ],
              )),
              const SizedBox(
                width: 300,
                height: 400,
                child: GlassyCard(
                    child: Center(
                  child: Text("This is another Glassy Card",
                      style: TextStyle(color: Colors.white)),
                )),
              ),
              Expanded(
                  child: Center(
                child: GlassyButton(
                    child: const Text("This is Glassy Button"),
                    onPressed: () {
                      showDialog(
                          context: context,
                          builder: ((context) {
                            return const GlassyDialog(
                                title: "This is Glassy Dialog Title",
                                child: Center(
                                  child: Text(
                                    "This is Glassy Dialog Content Text Widget",
                                    style: TextStyle(color: Colors.white),
                                  ),
                                ));
                          }));
                    }),
              ))
            ]),
          )
        ],
      ),
    );
  }
}

```

### List Of Widgets
* GlassyCard
* GlassyButton
* GlassyDialog


# License
Copyright (c) 2020 Sayed Mohammad Amin Emrani

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).

CHANGELOG.md
## [0.0.6] - July 17, 2022
* changes in README.md
## [0.0.5] - July 17, 2022
* fix the problems
## [0.0.4] - July 17, 2022
* fix the problems
## [0.0.3] - July 17, 2022
* initialize package


test/glassy_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:glassy/glassy.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


lib/glassy_card.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:glassy/glassy.dart';
import 'package:glassy/glassy_config.dart';

/// a card that has glass effects
class GlassyCard extends StatelessWidget {
  const GlassyCard({required this.child, this.config, super.key});

  /// a widget that placed as child of card
  final Widget child;

  /// optional configuration of glass effect
  final GlassyConfig? config;

  @override
  Widget build(BuildContext context) {
    return Card(
        color: (config ?? glassyConfig).backgroundColor,
        shape: RoundedRectangleBorder(
          side: BorderSide(
              width: 1, color: (config ?? glassyConfig).borderColor!),
          borderRadius: BorderRadius.circular((config ?? glassyConfig).radius),
        ),
        child: ClipRRect(
          borderRadius: BorderRadius.all(
              Radius.circular((config ?? glassyConfig).radius)),
          child: BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
              child: child),
        ));
  }
}


lib/glassy_button.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:glassy/glassy.dart';
import 'package:glassy/glassy_config.dart';

/// A Glassy Button that make a glass effect on elevated buttons
/// it can used withount GlassyConfig but is better to use with configs
class GlassyButton extends StatelessWidget {
  /// required params:
  ///   - a widget named child
  ///   - a void callback function named onPressed
  const GlassyButton(
      {Key? key,
      required this.onPressed,
      required this.child,
      this.config,
      this.padding = const EdgeInsets.all(15)})
      : super(key: key);

  /// The Callback Function named onPressed
  final VoidCallback? onPressed;

  /// The Child Widget Named child
  final Widget child;

  /// optional padding property
  final EdgeInsets padding;

  /// optional GlassyConfig property
  final GlassyConfig? config;

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      borderRadius:
          BorderRadius.all(Radius.circular((config ?? glassyConfig).radius)),
      child: BackdropFilter(
          filter: ImageFilter.blur(
            sigmaX: 10.0,
            sigmaY: 10.0,
          ),
          child: ElevatedButton(
              style: ElevatedButton.styleFrom(
                padding: padding,
                primary: (config ?? glassyConfig).backgroundColor,
                onPrimary: (config ?? glassyConfig).foregroundColor,
                shape: RoundedRectangleBorder(
                  side: BorderSide(
                      width: 1, color: (config ?? glassyConfig).borderColor!),
                  borderRadius:
                      BorderRadius.circular((config ?? glassyConfig).radius),
                ),
              ),
              onPressed: onPressed,
              child: child)),
    );
  }
}


lib/glassy_config.dart
import 'package:flutter/material.dart';

/// Glass Configuration that can determines configs of all glassy widgets
class GlassyConfig {
  GlassyConfig(
      {this.radius = 15,
      this.backgroundOpacity = 0.2,
      this.backgroundColor = Colors.grey,
      this.borderColor,
      this.foregroundColor = Colors.white,
      this.borderOpacity = 0.5}) {
    borderColor = borderColor ?? backgroundColor;

    borderColor = borderColor!.withOpacity(borderOpacity);
    backgroundColor = backgroundColor.withOpacity(backgroundOpacity);
  }

  /// the radius of all glassy widgets in px
  /// default value is "15"
  double radius;

  /// the opacity of background of all glassy widgets
  /// default value is "0.5"
  double backgroundOpacity;

  /// the opacity of border of all glassy widgets
  /// default value is "0.5"
  double borderOpacity;

  /// the color of background of all glassy widgets
  /// default value is "Colors.grey"
  Color backgroundColor;

  /// the color of foreground of all glassy widgets
  /// default value is "Colors.white"
  Color foregroundColor;

  /// the color of border of all glassy widgets
  /// default value is "Colors.grey"
  Color? borderColor;
}


lib/glassy_dialog.dart
import 'dart:math';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:glassy/glassy.dart';
import 'package:glassy/glassy_config.dart';

/// Glassy Dialog provides a dialog with Glass Effects
class GlassyDialog extends StatelessWidget {
  /// required params:
  ///   - child as a widget
  const GlassyDialog(
      {Key? key,
      required this.child,
      this.config,
      this.title,
      this.closeButton,
      this.leftBar,
      this.rightBar,
      this.size,
      this.onClose,
      this.fullScreen = false,
      this.direction = TextDirection.ltr})
      : super(key: key);

  /// size that determine dialog size
  final Size? size;

  /// dialog title that placed on center top of dialog
  final String? title;

  /// action buttons that placed on top left of dialog
  final List<Widget>? leftBar;

  /// action buttons that placed on top right of dialog
  final List<Widget>? rightBar;

  /// a boolean determine that dialog has close button or not
  final bool? closeButton;

  /// a widghet that placed in center of dialog as child
  final Widget child;

  /// a boolean determine that dialog is fullscreen or not
  final bool fullScreen;

  /// determine direction of dialog
  final TextDirection direction;

  /// optional configuration of glass effects
  final GlassyConfig? config;

  /// on dialog close callback
  final Function()? onClose;

  /// function that determines final dialog size
  Size dialogSize(BuildContext context) {
    if (size != null) {
      return size!;
    } else {
      return Size(min(MediaQuery.of(context).size.width * 0.95, 400),
          min(MediaQuery.of(context).size.width * 0.95, 400));
    }
  }

  @override
  Widget build(BuildContext context) {
    return ClipRect(
        child: BackdropFilter(
      filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
      child: Directionality(
          textDirection: TextDirection.rtl,
          child: OrientationBuilder(
            builder: (context, orientation) {
              return Dialog(
                insetPadding: fullScreen
                    ? EdgeInsets.zero
                    : const EdgeInsets.symmetric(
                        horizontal: 16.0, vertical: 24.0),
                backgroundColor: (config ?? glassyConfig).backgroundColor,
                shape: RoundedRectangleBorder(
                  side: BorderSide(
                      width: 1, color: (config ?? glassyConfig).borderColor!),
                  borderRadius: BorderRadius.circular(
                      fullScreen ? 0 : (config ?? glassyConfig).radius),
                ),
                child: Container(
                  width: dialogSize(context).width,
                  height: dialogSize(context).height == 0
                      ? null
                      : dialogSize(context).height,
                  padding: const EdgeInsets.all(0),
                  child: Column(
                    children: [
                      Row(
                        children: [
                          if (closeButton == null || closeButton == true)
                            IconButton(
                              icon: Icon(
                                Icons.clear,
                                color: (config ?? glassyConfig).foregroundColor,
                              ),
                              onPressed: () {
                                Navigator.of(context).pop();
                                if (onClose != null) {
                                  onClose!();
                                }
                              },
                            ),
                          if (rightBar != null)
                            Row(
                              children: rightBar!,
                            ),
                          Expanded(
                            child: Padding(
                              padding: const EdgeInsets.all(15),
                              child: FittedBox(
                                fit: BoxFit.scaleDown,
                                child: Text(
                                  title == null ? " " : title!,
                                  style: TextStyle(
                                      fontSize: 20,
                                      color: (config ?? glassyConfig)
                                          .foregroundColor),
                                ),
                              ),
                            ),
                          ),
                          if (leftBar != null)
                            Row(
                              children: leftBar!,
                            ),
                        ],
                      ),
                      Expanded(child: child)
                    ],
                  ),
                ),
              );
            },
          )),
    ));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:glassy/glassy.dart';
import 'package:glassy/glassy_button.dart';
import 'package:glassy/glassy_card.dart';
import 'package:glassy/glassy_config.dart';
import 'package:glassy/glassy_dialog.dart';

void main() {
  Glassy().setConfig(GlassyConfig(
      radius: 15,
      backgroundColor: Colors.grey,
      backgroundOpacity: 0.2,
      borderOpacity: 0.5));
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Glassy Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const DemoPage(),
    );
  }
}

class DemoPage extends StatelessWidget {
  const DemoPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        fit: StackFit.expand,
        children: [
          Image.asset("assets/images/test.jpg", fit: BoxFit.cover),
          SizedBox(
            child: Column(children: [
              GlassyCard(
                  child: Row(
                children: [
                  IconButton(
                      onPressed: () {},
                      icon: const Icon(
                        Icons.search,
                        color: Colors.white,
                      )),
                  const Expanded(
                      child: Text(
                    "This is a GlassyCard",
                    textAlign: TextAlign.center,
                    style: TextStyle(
                      color: Colors.white,
                    ),
                  ))
                ],
              )),
              const SizedBox(
                width: 300,
                height: 400,
                child: GlassyCard(
                    child: Center(
                  child: Text("This is another Glassy Card",
                      style: TextStyle(color: Colors.white)),
                )),
              ),
              Expanded(
                  child: Center(
                child: GlassyButton(
                    child: const Text("This is Glassy Button"),
                    onPressed: () {
                      showDialog(
                          context: context,
                          builder: ((context) {
                            return const GlassyDialog(
                                title: "This is Glassy Dialog Title",
                                child: Center(
                                  child: Text(
                                    "This is Glassy Dialog Content Text Widget",
                                    style: TextStyle(color: Colors.white),
                                  ),
                                ));
                          }));
                    }),
              ))
            ]),
          )
        ],
      ),
    );
  }
}
